nupic/research/frameworks/vernon/handlers.py

```diff
@@ -1311,15 +1313,20 @@ def update_params(cls, params, loss, lr, distributed=False):
                 grad.data /= size
 
         if gradients is not None:
-            params = list(params)
-            for p, g in zip(params, gradients):
+            for g, (name, p) in zip(gradients, named_params.items()):
                 if g is not None:
-                    p.add_(g, alpha=-lr)
+                    updated = p.add(g, alpha=-lr)
+
+                    # Update in place inn a way that preserves grads.
+                    parent_module = get_parent_module(model, name)
+                    base_name = name.split(".")[-1]
+                    parent_module._parameters[base_name] = updated
```

Comment:
I don't understand the bug, why can't it be done inplace?

Reply:
This is a good question and it deserves a good answer. I'm still trying to understand a bit more about why it's needed (other than the fact it fixed a bug PyTorch was throwing). Let me get back to you on this.